integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-spark-service-bundle/nifi-spark-service/src/main/java/com/thinkbiganalytics/nifi/v2/spark/SparkJobserverService.java

```diff
@@ -0,0 +1,357 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+@Tags({"thinkbig", "spark"})
+@CapabilityDescription("Provides long running spark contexts and shared RDDs using Spark jobserver.")
+public class SparkJobserverService extends AbstractControllerService implements JobService {
+
+    /**
+     * A property to get the Spark Jobserver connection URL used to connect to a jobserver.
+     */
+    public static final PropertyDescriptor JOBSERVER_URL = new PropertyDescriptor.Builder()
+        .name("Jobserver URL")
+        .description("A URL used to connect to Spark jobserver.")
+        .defaultValue(null)
+        .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+        .required(true)
+        .build();
+
+    /**
+     * A property to set the default timeout for Sync Spark Jobserver calls
+     */
+    public static final PropertyDescriptor SYNC_TIMEOUT = new PropertyDescriptor.Builder()
+        .name("Sync Timeout")
+        .description("Number of seconds before timeout for Sync Spark Jobserver calls.")
+        .defaultValue("6000")
+        .addValidator(StandardValidators.LONG_VALIDATOR)
+        .required(true)
+        .build();
+
+    /**
+     * List of properties
+     */
+    private List<PropertyDescriptor> properties;
+    private volatile ISparkJobServerClient client;
+    private volatile String jobServerUrl;
+    private volatile String syncTimeout;
+    private volatile Timer contextTimeoutTimer = new Timer();
+    private volatile Map<String, SparkContextState> sparkContextsStates = Collections.synchronizedMap(new HashMap<String, SparkContextState>());
+
+    @Override
+    protected void init(@Nonnull final ControllerServiceInitializationContext config) throws InitializationException {
+
+        // Create list of properties
+        final List<PropertyDescriptor> props = new ArrayList<>();
+        props.add(JOBSERVER_URL);
+        props.add(SYNC_TIMEOUT);
+        properties = Collections.unmodifiableList(props);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    /**
+     * Configures Spark Jobserver client by creating an instance of the
+     * {@link ISparkJobServerClient} based on configuration provided with
+     * {@link ConfigurationContext}.
+     * <p>
+     * This operation makes no guarantees that the actual connection could be
+     * made since the underlying system may still go off-line during normal
+     * operation of the service.
+     *
+     * A Context Timeout task is also created to delete any contexts which have
+     * been inactive for some time.
+     *
+     * @param context the configuration context
+     * @throws InitializationException if unable to create a Spark Jobserver connection
+     */
+    @OnEnabled
+    public void onConfigured(final ConfigurationContext context) throws InitializationException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'org.apache.nifi.reporting.InitializationException', as it cannot be thrown from method's body. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)